### ISSUE TITLE
multinode ppo training extremely slow

### ISSUE THREAD

**Issue by babu111** [OpenRLHF/OpenRLHF#382]
Issue: When I run ppo with a single node, the train step finishes in dozens of seconds. However, after using 3 nodes, the train step takes half an hour to finish.

Now I describe how I setup multi-node Ray.
To activate the head node, I run `ray start --head --port=6379`
After head node is activated, I find its ip in the local network, such as 10.2.1.121.

To activate the worker nodes, I ssh into each machine, then run
`ray start --address='10.2.1.121:6379'`

The batch size I use for single-node is 
``
   --train_batch_size 32 \
   --micro_train_batch_size 2 \
   --rollout_batch_size 64 \
   --micro_rollout_batch_size 8 \
``
The batch size I use for multi-node is 
``
   --train_batch_size 72 \
   --micro_train_batch_size 2 \
   --rollout_batch_size 792 \
   --micro_rollout_batch_size 4 \
``
The env I use is recommended by the latest OpenRLHF, so it should use NCCL as backend? I'm not sure about this.

**Comment by hijkzzz** [comment]
disable adam offload or keep the model in one node.

**Comment by babu111** [comment]
Actually, rollout is quite fast. I rollout 792 samples in just a few min.
I find that the issue of slowness is due to slow comm between nodes.
when the training step is stuck, I visualize the status using ray.dashboard. It shows that the speed of comm between nodes is 11.73MB/s, which is too slow for nccl. So I'm suspecting the nccl backend is not being used. But I don't know how to check if it's true, or how to make sure nccl is activated.

**Comment by babu111** [comment]
I'm not using adam offload 
From dashboard, each model is in one node.
I have 6 actors in node 0. 
6 ref actors in node 1. 
1 reward actor in node 0.
the rest are vllm engines

**Comment by babu111** [comment]
could you share how you setup the multinode to train llama3 70B? like the cluster setup and how ray is configured?
I think it's most likely to be a problem with the comms, so I'm asking my gpu provider for hardware details also.

**Comment by babu111** [comment]
Also, in the code, we create several torch process groups. These groups contain remote actors or vllm engines.
When we build the process groups and specify nccl as backend, how does ray know the comm specifics to actually use nccl as backend? I'm suspecting that it's using ethernet for comms.

**Comment by hijkzzz** [comment]
> Also, in the code, we create several torch process groups. These groups contain remote actors or vllm engines. When we build the process groups and specify nccl as backend, how does ray know the comm specifics to actually use nccl as backend? I'm suspecting that it's using ethernet for comms.

We have tested on both 16 A100s (2 nodes) and 32 A100s (4 nodes), and the 70B model works fine. Of course, our cluster is in an NCCL environment.

**Comment by babu111** [comment]
Yeah, I've checked this file before. I almost got to a point where I know every file in the OpenRLHF repo haha.
What I want to ask about is how the ray cluster is set up, and the details of how torch process groups work with ray remotes.

**Comment by hijkzzz** [comment]
Ray is merely a glue/process launcher and does not affect or participate in DeepSpeed's NCCL communication.
We have provided the ray launch script on slurm cluster:
https://github.com/OpenRLHF/OpenRLHF/blob/main/examples/scripts/train_ppo_llama_ray_slurm.sh

**Comment by babu111** [comment]
Hmm, I see, I'm using ray start and ray job submit also.
I printed out the backend used in the process groups, and it's nccl.
But I still have no idea why node to node comm is that slow with nccl.

**Comment by hijkzzz** [comment]
This is very strange. I haven't encountered a similar issue before. Ray doesn't affect whether it's single-node or multi-node communication, as Ray is only responsible for launching DeepSpeed processes on different machines. Then DeepSpeed handles the communication on its own. We've only introduced one additional communication, which is using NCCL to synchronize weights from DeepSpeed to VLLM.

**Comment by babu111** [comment]
Yeah, I used grafana to visualize the results. I verified that the training is indeed stuck at broadcasting new weight to vllm engines (across nodes)

**Comment by babu111** [comment]
<img width="526" alt="image" src="https://github.com/user-attachments/assets/4bd922b3-674d-48a7-ac12-4c3c7c4e8239">

**Comment by babu111** [comment]
<img width="544" alt="image" src="https://github.com/user-attachments/assets/6b187001-d698-48b6-a9b9-e99139a647b6">
this is with vllm_tensor_parallel_size 1

**Comment by hijkzzz** [comment]
This is likely due to the failure in establishing the NCCL group:
https://github.com/OpenRLHF/OpenRLHF/blob/6a13d95439746251c3ae975b5d69e39c14ea614c/openrlhf/trainer/ray/ppo_actor.py#L99
then the broadcast hang?
Could you try vLLM 0.4.2 or user `--vllm_sync_backend gloo`?
